Exporter: allow extrapolation into excisions

nilsvu opened the pull request:
## Proposed changes

This is needed to fill the excision region with (non-constraint-satisfying) data so it can be imported into moving puncture codes.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add instantiations for CartesianToSpherical

wthrowe commented on src/Domain/BlockLogicalCoordinates.cpp:
```diff
@@ -125,14 +126,57 @@ block_logical_coordinates_single_point(
+    const double radial_distance_this_block =
+        x_logical->get(direction.dimension()) * direction.sign();
+    if (radial_distance_this_block < 1.) {
+      continue;
```
If I follow the logic correctly, if we hit this case the point is not inside the excision sphere.  Should that be an error or an early return?

wthrowe commented on src/Domain/BlockLogicalCoordinates.hpp:
```diff
@@ -69,10 +71,30 @@ auto block_logical_coordinates(
+ * The point will be reported to belong to one of the blocks that abut the
+ * excision sphere and will have a logical coordinate outside the range [-1, 1]
+ * in the radial direction. This is useful for extrapolating data into excision
+ * regions.
```
What happens if `input_point` is not inside `excision_sphere`?  Either document the result or state it is considered invalid to call the function in that case.

wthrowe commented on src/Domain/BlockLogicalCoordinates.cpp:
```diff
@@ -125,14 +126,57 @@ block_logical_coordinates_single_point(
+      continue;
+    }
+    // The checks above should leave only 1 valid block, so return that
+    return make_id_pair(domain::BlockId(block_id),
```
Could there be two valid blocks for points on the extrapolation of the boundary between two blocks into the sphere?  The blocks are being pulled from an unordered container, so always choosing the first one isn't deterministic.

wthrowe commented on src/IO/Exporter/Exporter.cpp:
```diff
@@ -123,7 +125,222 @@ void interpolate_to_points(
+    for (size_t s = 0; s < block_logical_coords->size(); ++s) {
+      auto& block_logical_coord = (*block_logical_coords)[s];
+      // Process only points inside the excision sphere. These points were found
+      // by `block_logical_coordinates` above, have a radial logical coordinate
```
`block_logical_coordinates` doesn't appear anywhere above.

wthrowe commented on src/IO/Exporter/Exporter.hpp:
```diff
@@ -45,6 +45,15 @@ struct ObservationStep {
+ * \cite Etienne2008re adjusted for distorted excisions: we choose uniformly
+ * spaced radial anchor points spaced as $\Delta r = 0.3 r_\mathrm{AH}$ in the
+ * grid frame (where the excision is spherical), then map the anchor points to
+ * the distorted frame (where we have the target point) and do a 7th order
```
The code has `constexpr size_t extrapolation_order = 8;`

wthrowe reviewed (changes requested):
Exporter test failed on one of the CI jobs.

nilsvu commented:
Thanks for your review @wthrowe! I pushed a fixup.

nilsvu commented:
I'm investigating the NaNs in some of the unit tests, but the fixup is ready for your review @wthrowe

wthrowe commented on src/Domain/BlockLogicalCoordinates.cpp:
```diff
@@ -153,22 +153,26 @@ BlockLogicalCoords<Dim> block_logical_coordinates_in_excision(
-    // range [-1, 1]
+    // range [-1, 1]. Also discard if the point is on the upper angular face of
+    // the block, to disambiguate which of the two blocks sharing the face the
+    // point is in.
```
This doesn't work.  The blocks don't have aligned frames, so some interfaces are two lower sides or two upper sides.

wthrowe commented on src/IO/Exporter/Exporter.hpp:
```diff
@@ -49,11 +49,11 @@ struct ObservationStep {
- * the distorted frame (where we have the target point) and do a 7th order
- * polynomial extrapolation into the excision region.
+ * the distorted frame (where we have the target point) and do a polynomial
+ * extrapolation into the excision region.
```
Is it 8th order or not?

nilsvu replied:
It's a 7th order polynomial with 8 collocation points. However in the `polynomial_interpolation` function the name `order` is `degree + 1`, so I used that naming here as well:
https://github.com/sxs-collaboration/spectre/blob/bc197745bb08b2397fb31dda402a10b4d17cc856/src/NumericalAlgorithms/Interpolation/PolynomialInterpolation.cpp#L19
To avoid confusion I renamed to `num_extrapolation_anchors`.

nilsvu committed:
Exporter: allow extrapolation into excisions

This can be useful to fill the excision region with
(non-constraint-satisfying) data so it can be imported into
moving puncture codes.

nilsvu commented:
@wthrowe I pushed a new fixup. I dropped the changes to `BlockLogicalCoordinates.*pp` and instead determine if the target point is inside an excision in the distorted frame. I think this is more robust and addresses your comments above.

wthrowe commented:
OK, looks good.  Squash and fix the clang-tidy complaint.

wthrowe commented:
@nilsvu I filed a CI failure as #6168 and I'm rerunning it as I'm guessing it's unrelated to this change.  Do you agree with that guess?

Edit: (The gcc-11 Release failure is obviously unrelated so I'm ignoring it.)

nilsvu commented:
I don't see a connection 🤷‍♂️

wthrowe reviewed (approved):
Sounds good.  Let's let the rerun finish, but feel free to merge once that's done (or I'll do it later, but I'm leaving for the day soon).
